r-package/grf/R/survival_forest.R

```diff
@@ -0,0 +1,206 @@
+#'                        further be cut by a factor of honesty.fraction. Default is 0.5.
+#' @param mtry Number of variables tried for each split. Default is
+#'             \eqn{\sqrt p + 20} where p is the number of variables.
+#' @param min.node.size A target for the minimum number of observations in each tree leaf. Note that nodes
+#'                      with size smaller than min.node.size can occur, as in the original randomForest package.
+#'                      Default is 5.
+#' @param honesty Whether to use honest splitting (i.e., sub-sample splitting). Default is TRUE.
+#'  For a detailed description of honesty, honesty.fraction, honesty.prune.leaves, and recommendations for
+#'  parameter tuning, see the grf
+#'  \href{https://grf-labs.github.io/grf/REFERENCE.html#honesty-honesty-fraction-honesty-prune-leaves}{algorithm reference}.
+#' @param honesty.fraction The fraction of data that will be used for determining splits if honesty = TRUE. Corresponds
+#'                         to set J1 in the notation of the paper. Default is 0.5 (i.e. half of the data is used for
+#'                         determining splits).
+#' @param honesty.prune.leaves If TRUE, prunes the estimation sample tree such that no leaves
+#'  are empty. If FALSE, keep the same tree as determined in the splits sample (if an empty leave is encountered, that
+#'  tree is skipped and does not contribute to the estimate). Setting this to FALSE may improve performance on
+#'  small/marginally powered data, but requires more trees (note: tuning does not adjust the number of trees).
+#'  Only applies if honesty is enabled. Default is TRUE.
+#' @param alpha A tuning parameter that controls the maximum imbalance of a split. Default is 0.05
+#'  (meaning the count of failures on each side of a split has to be at least 5 % of the total observation count in a node)
+#' @param compute.oob.predictions Whether OOB predictions on training set should be precomputed. Default is TRUE.
+#' @param num.threads Number of threads used in training. By default, the number of threads is set
+#'                    to the maximum hardware concurrency.
+#' @param seed The seed of the C++ random number generator.
+#'
+#' @return A trained survival_forest forest object. The attribute `failure.times` contains the unique failure
+#'  times in the data set.
+#'
+#' @examples
+#' \dontrun{
+#' # Train a standard survival forest.
+#' n <- 100
+#' p <- 5
+#' X <- matrix(rnorm(n * p), n, p)
+#' latent.time <- -log(runif(n)) * exp(0.1 * X[, 1])
+#' censor.time <- rexp(n)
+#' Y <- pmin(latent.time, censor.time)
+#' D <- as.integer(latent.time <= censor.time)
+#' s.forest <- survival_forest(X, Y, D)
+#'
+#' # Predict using the forest.
+#' X.test <- matrix(0, 101, p)
+#' X.test[, 1] <- seq(-2, 2, length.out = 101)
+#' s.pred <- predict(s.forest, X.test)
+#'
+#' # Predict on out-of-bag training samples.
+#' s.pred <- predict(s.forest)
+#' }
+#'
+#' @export
+survival_forest <- function(X, Y, D,
```

Comment:
For performance, it'd be helpful if the user could provide a vector of event times at which they want to fit the survival curve; then we could round down the actual even times to the nearest event time in this vector. Otherwise, when we're running this with a sample size of, say, `n = 10000`, it seems like we're forcing the forest to learn the survival function at a needlessly fine granularity.

Reply:
I thought the user could just round the `Y` vector, but this makes much more sense. It is added in commit 70a0e5d

Reply:
An example (from the docstring):

```R
# Train a standard survival forest.
n <- 2000
p <- 5
X <- matrix(rnorm(n * p), n, p)
failure.time <- exp(0.5 * X[, 1]) * rexp(n)
censor.time <- 2 * rexp(n)
Y <- pmin(failure.time, censor.time)
D <- as.integer(failure.time <= censor.time)
s.forest <- survival_forest(X, Y, D)

# Predict on out-of-bag training samples.
s.pred <- predict(s.forest)

# Plot the survival curve for the first five individuals.
matplot(s.pred$failure.times, t(s.pred$predictions[1:5, ]),
        xlab = "failure time", ylab = "survival function (OOB)",
        type = "l", lty = 1)

# Train the forest on a less granular grid.
failure.summary <- summary(Y[D == 1])
events <- seq(failure.summary["Min."], failure.summary["Max."], by = 0.1)
s.forest.grid <- survival_forest(X, Y, D, failure.times = events)
s.pred.grid <- predict(s.forest.grid)
matpoints(s.pred.grid$failure.times, t(s.pred.grid$predictions[1:5, ]),
          type = "l", lty = 2)
```

![Screen Shot 2020-04-29 at 10 26 03](https://user-images.githubusercontent.com/7185264/80626852-e33a3880-8a03-11ea-89bb-4ca4d2f96419.png)